### PR TITLE
Verify SQL MERGE access control checks

### PR DESCRIPTION
**Pull request by djsstarburst**
## Description

This commit adds a MERGE test in BaseConnectorTest that verifies
that in order execute a MERGE, you must have appropriate access
control rights.  Specifically, if there is a source table, you must
have SELECT on the table columns referenced; if the MERGE does
inserts, you must have INSERT on the target table; if the MERGE
does deletes you must have DELETE on the target table; and if
the MERGE does updates, you must have UPDATE on the target
table.

> Is this change a fix, improvement, new feature, refactoring, or other?

It is an added test.

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

None of the above.

> How would you describe this change to a non-technical end user or system administrator?

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(x) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java
Comment (dain) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:4365]: I think the formatting is off here

**Review by dain — APPROVED** [review]
Other than the formatting issue, looks good

**Review by electrum — CHANGES_REQUESTED** [review]
These should move to `TestAccessControl`, as discussed offline. We don't need test data since we are only testing access control checks, not the merge functionality. I created #13665 which allows testing the positive case.

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java
Comment (djsstarburst) [reply testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java]: Fixed.

**Comment by djsstarburst** [comment]
@electrum suggested that I move this test to `TestAccessControl`.  However, that test uses the Blackhole connector, which didn't support MERGE.

I added another commit to the start of my PR that moves the MERGE access control checks up to the top of `StatementAnalysis.visitMerge`. It a good change, it's not sufficient to move the test to `TestAccessControl`.  The problem is that we don't know which if any columns of the source and target tables until `StatementAnalysis.visitTable` is called, and it make a call to get the merge rowId handle.

@electrum produced an ersatz implementation of MERGE for Blackhole, and maybe that makes sense as a followup.  But at this point, I’m mildly in favor of merging my existing PR, now with the commit that does access checks earlier.

**Comment by djsstarburst** [comment]
> I created https://github.com/trinodb/trino/pull/13665 which allows testing the positive case.

I don't see how that PR can work, because the body of `storeMergedRows` is empty.

**Comment by electrum** [comment]
It works because the Black Hole connector doesn't store or produce data (unless you ask it to produce nulls), so there is nothing to match or update. You can see that `BlackHolePageSink` ignores the provided page.

**Comment by djsstarburst** [comment]
I rebased on top of tip master, which has @electrum's ersatz implementation of MERGE for the Blackhole connector.  That let me remove the test from `BaseConnectorTest` and move it to `TestAccessControl`, after I removed the final successful MERGE operation and verification, which Blackhole can't execute.

I think this now satisfies the concerns raised, and is ready to merge.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java
Comment (electrum) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:2847]: Nit: we can use `HashSet` here with the variable named `allUpdateColumnNames` since we are just building this to pass to a function. Using an immutable builder is overkill and makes the code harder to read.

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java
Comment (electrum) [inline testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:402]: No need to insert data, since we aren't testing the actual merge functionality, and the blackhole connector will discard the data anyway. (so it's confusing to the reader as to why we do it)

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java
Comment (electrum) [inline testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:458]: We shouldn't need this method

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java
Comment (electrum) [inline testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:422]: Use `assertAccessDenied`, like the other tests in this class.

**Review by electrum — APPROVED** [review]
Looks good overall. A few minor comments. Thanks for updating this.

**Comment by electrum** [comment]
Blackhole should be able to run the final `MERGE` using `assertAccessAllowed`. The existing one that this PR replaces runs.

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java
Comment (djsstarburst) [reply testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java]: Removed.

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java
Comment (djsstarburst) [reply testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java]: Changed.

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java
Comment (djsstarburst) [reply testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java]: Removed.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java
Comment (djsstarburst) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java]: Changed.

**Comment by djsstarburst** [comment]
> Blackhole should be able to run the final MERGE using assertAccessAllowed. The existing merge test using assertAccessAllowed that this PR replaces runs.

Added.

**Comment by electrum** [comment]
Thanks!

**Pull request merged**
